subsys/usb/class/audio/audio.c

```diff
@@ -0,0 +1,958 @@
+		}
+
+/**
+ * Helper function for getting channel number directly from the
+ * feature unit descriptor.
+ */
+static u8_t get_num_of_channels(struct feature_unit_descriptor *fu)
+{
+	return (fu->bLength - FU_FIXED_ELEMS_SIZE)/sizeof(u16_t);
+}
+
+/**
+ * Helper function for getting supported controls directly from
+ * the feature unit descriptor.
+ */
+static u16_t get_controls(struct feature_unit_descriptor *fu)
+{
+	return *(u16_t *)((u8_t *)fu + BMA_CONTROLS_OFFSET);
+}
+
+/**
+ * Helper function for getting the device streaming direction
+ */
+static enum usb_audio_direction get_fu_dir(struct feature_unit_descriptor *fu)
+{
+	struct output_terminal_descriptor *ot  =
+	(struct output_terminal_descriptor *)((u8_t *)fu + fu->bLength);
+	enum usb_audio_direction dir;
+
+	if (ot->wTerminalType == USB_AUDIO_USB_STREAMING) {
+		dir = USB_AUDIO_IN;
+	} else {
+		dir = USB_AUDIO_OUT;
+	}
+
+	return dir;
+}
+
+/**
+ * Helper function for fixing controls in feature units descriptors.
+ */
+static void fix_fu_descriptors(struct usb_if_descriptor *iface)
+{
+	struct cs_ac_interface_descriptor_header *header;
+	struct feature_unit_descriptor *fu;
+
+	header = (struct cs_ac_interface_descriptor_header *)
+			((u8_t *)iface + USB_PASSIVE_IF_DESC_SIZE);
+
+	fu = (struct feature_unit_descriptor *)((u8_t *)header +
+						header->bLength +
+						INPUT_TERMINAL_DESC_SIZE);
+
+	/* start from 1 as elem 0 is filled when descriptor is declared */
+	for (int i = 1; i < get_num_of_channels(fu); i++) {
+		*(fu->bmaControls + i) = fu->bmaControls[0];
+	}
+
+	if (header->bInCollection == 2) {
+		fu = (struct feature_unit_descriptor *)((u8_t *)fu +
+			fu->bLength +
+			INPUT_TERMINAL_DESC_SIZE +
+			OUTPUT_TERMINAL_DESC_SIZE);
+		for (int i = 1; i < get_num_of_channels(fu); i++) {
+			*(fu->bmaControls + i) = fu->bmaControls[0];
+		}
+	}
+}
+
+/**
+ * Helper function for getting pointer to feature unit descriptor.
+ * This is needed in order to address audio specific requests to proper
+ * controls struct.
+ */
+static struct feature_unit_descriptor *get_feature_unit(
+				struct usb_audio_dev_data_t *audio_dev_data,
+				u8_t *device, u8_t fu_id)
+{
+	struct feature_unit_descriptor *fu;
+
+	fu = (struct feature_unit_descriptor *)
+		((u8_t *)audio_dev_data->header_descr +
+		audio_dev_data->header_descr->bLength +
+		INPUT_TERMINAL_DESC_SIZE);
+
+	if (fu->bUnitID == fu_id) {
+		*device = 0;
+		return fu;
+	}
+	/* skip to the next Feature Unit */
+	fu = (struct feature_unit_descriptor *)
+			((u8_t *)fu + fu->bLength +
+			INPUT_TERMINAL_DESC_SIZE +
+			OUTPUT_TERMINAL_DESC_SIZE);
+	*device = 1;
+
+	return fu;
+}
+
+/**
+ * @brief This is a helper function user to inform the user about
+ * possibility to write the data to the device.
+ */
+static void audio_dc_sof(struct usb_cfg_data *cfg,
+			 struct usb_audio_dev_data_t *dev_data)
+{
+	u8_t ep_addr;
+
+	for (int i = 0; i < cfg->num_endpoints; i++) {
+		ep_addr = cfg->endpoint[0].ep_addr;
```

Comment:
Something is probably wrong here. There is a loop but in each iteration it only checks the first endpoint.

Reply:
Ah, yea... Initial thought was it should loop over endpoints but then I made IN endpoints always 1st. I will change this -> remove the loop.